src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,467 @@
+        {
+            final Collection<Pair<Segment, Segment>> segs = getIntersectingSegments(linePolyline,
+                    lineItem.asPolyLine());
+            final Segment max = segs.stream().map(Pair::getKey).distinct().max(segmentComparator)
+                    .orElse(null);
+            final Collection<Segment> crosses = segs.stream().filter(p -> p.getLeft().equals(max))
+                    .map(Pair::getValue).collect(Collectors.toSet());
+            final PolyLine coast = new PolyLine(crosses.stream()
+                    .flatMap(c -> Stream.of(c.first(), c.last())).toArray(Location[]::new));
+            if (isRightOf(coast, linePolyline.last())
+                    || lineItem.asPolyLine().contains(linePolyline.last()))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if a line ends in a waterway sink
+     *
+     * @param lineItem
+     *            The LineItem to check
+     * @return {@code true} if the waterway can reasonably be expected to end
+     */
+    public boolean doesLineEndInSink(final LineItem lineItem)
+    {
+        // If the way is a sink, it obviously ends in a sink...
+        if (this.waterwaySinkTagFilter.test(lineItem))
+        {
+            return true;
+        }
+        if (lineItem instanceof Line)
+        {
+            final Line line = (Line) lineItem;
+            final Location last = line.asPolyLine().last();
+            final Set<LocationItem> one = Iterables.stream(line.getAtlas().nodesAt(last))
+                    .filter(LocationItem.class::isInstance).map(LocationItem.class::cast)
+                    .collectToSet();
+            final Set<LocationItem> two = Iterables.stream(line.getAtlas().pointsAt(last))
+                    .filter(LocationItem.class::isInstance).map(LocationItem.class::cast)
+                    .collectToSet();
+            if (Stream.concat(one.stream(), two.stream())
+                    .anyMatch(this.waterwaySinkTagFilter::test))
+            {
+                return true;
+            }
+        }
+        else if (lineItem instanceof Edge)
+        {
+            return this.waterwaySinkTagFilter.test(((Edge) lineItem).end());
+        }
+        return lineItem.getAtlas()
+                .areasCovering(lineItem.asPolyLine().last(), this.waterwaySinkTagFilter::test)
+                .iterator().hasNext();
+    }
+
+    /**
+     * Check if the last location on a line also happens to be connected to a waterway.
+     *
+     * @param line
+     *            The line to check
+     * @return {@code true} if the line ends on a waterway (the last node is part of a waterway)
+     */
+    public boolean doesLineEndOnWaterway(final LineItem line)
+    {
+        final List<LineItem> waterways = new ArrayList<>();
+        line.getAtlas()
+                .lineItemsIntersecting(line.asPolyLine().last().boxAround(Distance.ONE_METER),
```

Comment:
I think you can simplify this method by using `Atlas.lineItemsContaining(Location)`, and filtering from there.

Reply:
I didn't see that when I was coding. I probably similar code in the other water check PR I made.

Thanks. :)